WASAPI code improvements

CookiePLMonster opened the pull request:
This PR brings numerous improvements to WASAPI code in Dolphin, improving code readability as well as fixing numerous issues.

As a preparation, [WIL](https://github.com/microsoft/wil) has been added, since it's very useful for ensuring proper RAII-based destruction for several WinAPI types - in my case it was proper scoped calls to `CoUninitialize` and proper destruction of previously leaking (in some cases) `PROPVARIANT`.

All COM objects have been moved to use WRL, which ensures their proper release. Previously, there were numerous code paths where functions would return without cleaning up their COM objects, effectively leaking references.

`device->GetId(&device_id);` was present in several places, never using the return value nor freeing it - leaking memory for no reason.

Thread synchronization model for WASAPI Handler thread has been simplified by not detaching said thread. Not only it improved code readability, but also fixed a spinlock occuring if `WASAPIStream::SetRunning(true)` failed in **any** way - a following call to `WASAPIStream::SetRunning(false)` would then spin endlessly checking for `m_stopped`. Now this boolean is gone, and so is the problem.

`std::string_view` is now used where applicable. I opted for constructing a `std::string` out of `std::string_view` in `HandleWinAPI`, because it happens in code path which is taken relatively rarely - in a perfect case it's never taken.

I modified `SoundLoop` not to do volume adjustment unless absolutely necessary, saving on a relatively costly loop. For 100% volume, volume adjustment is not needed, and for 0% volume a `AUDCLNT_BUFFERFLAGS_SILENT` is now specified instead of zeroing the buffer. I also removed several nullptr checks which were not thread safe anyway, so only brought a false sense of safety.

CookiePLMonster commented:
Build fails because VS2017 on the build server is too old. Seems to be the same error as reported here:
https://github.com/microsoft/wil/issues/25

Apparently it can be worked around by defining `__WI_LIBCPP_HAS_NO_IS_AGGREGATE`, but I guess upgrading VS2017 on the build server (if possible) is better?

Warepire commented on Source/Core/AudioCommon/WASAPIStream.cpp:
```diff
@@ -82,134 +79,115 @@ static bool HandleWinAPI(std::string message, HRESULT result)
 }
 
-IMMDevice* WASAPIStream::GetDeviceByName(std::string name)
+ComPtr<IMMDevice> WASAPIStream::GetDeviceByName(std::string name)
```
`GetDeviceByName()` and `GetAvailableDevices()` do a lot of the same stuff. Could some of that be broken out into a separate method / function?

CookiePLMonster replied:
I'll look into this, as effectively only contents of `for (u32 i = 0; i < count; i++)` loop differ...

Warepire commented on Source/Core/AudioCommon/WASAPIStream.cpp:
```diff
@@ -82,134 +79,115 @@ static bool HandleWinAPI(std::string message, HRESULT result)
+ComPtr<IMMDevice> WASAPIStream::GetDeviceByName(std::string name)
 {
-  CoInitialize(nullptr);
+  HRESULT result = CoInitializeEx(nullptr, COINIT_MULTITHREADED);
```
This in particular is also done in the constructor, is there a reason why you do it again?

CookiePLMonster replied:
No, no reason here, I guess it's a result of copypasting things around from `GetAvailableDevices`. Although I'll oppose removing it, since `GetDeviceByName` is a static public function, so in principle there is nothing preventing it from being called outside the constructor/init logic - unless we want to make it private instead.

Warepire replied:
I completely overlooked they were static, in that sense it makes perfect sense.

lioncash commented:
The AudioCommon CMakeLists file should also be modified to handle the use of WIL.

CookiePLMonster commented:
> The AudioCommon CMakeLists file should also be modified to handle the use of WIL.

I normally steer clear of CMake, so will
```
target_compile_definitions(audiocommon PRIVATE WIL_SUPPRESS_EXCEPTIONS)
```

be valid syntax? The other part about adding `WIL\include` to include paths seems easy enough.

BhaaLseN commented on Source/Core/AudioCommon/WASAPIStream.h:
```diff
@@ -35,18 +38,21 @@ class WASAPIStream final : public SoundStream
-  IAudioRenderClient* m_audio_renderer = nullptr;
-  IMMDeviceEnumerator* m_enumerator = nullptr;
-  HANDLE m_need_data_event = nullptr;
+  // This MUST be located before ComPtr's underneath!
```
Could you please add more detail to this comment? Such as "must be located before...to ensure CoInitialize is called/COM stuff works/for reasons" or whatever makes most sense to describe the _why_.

BhaaLseN commented on Source/Core/AudioCommon/WASAPIStream.cpp:
```diff
@@ -39,177 +44,121 @@ WASAPIStream::WASAPIStream()
+
+  wil::unique_couninitialize_call cleanup;
+  if (FAILED(result))
+    cleanup.release();  // Only call CoUninitialize on success
```
Kinda same reasoning as for the "It's fine" comment: _why_? Omitting that just leaves the reader with more questions than answers (at least in my case where I'm fairly rusty in COM and stuff)

CookiePLMonster commented on Source/Core/AudioCommon/WASAPIStream.h:
```diff
@@ -8,6 +8,9 @@
 // clang-format off
 #include <Windows.h>
 #include <mmreg.h>
+#include <objbase.h>
```
Actually they do. `wil/resource.h` **must** be included after objbase.h, so it may be better to keep it here. Or move it outside since alphabetical ordering happens to be correct?

CookiePLMonster replied:
I would keep it inside this block to make it clear ordering is important. However, I've moved WRL include away to a separate block because that does not need to follow a specific ordering.

BhaaLseN replied:
If ordering does matter, leave it in there. I just want to make sure this isn't mixed when it could just be its own sorted block instead.

CookiePLMonster commented on Source/Core/AudioCommon/WASAPIStream.cpp:
```diff
@@ -13,17 +13,22 @@
 #include <devpkey.h>
 #include <functiondiscoverykeys_devpkey.h>
 #include <thread>
+#include <wil/resource.h>
```
`thread` doesn't need to be there indeed and can be moved away, but WIL must be included after any things you want to wrap it with - so I would keep it there to signify that order of includes is important, even though alphabetical ordering would incidentally be "correct" here.

CookiePLMonster commented:
I have fixed CMake definitions (and tested them), also reworded all mentioned comments. Also I added one more commits, because I just noticed a `std::thread` constructor with invoke semantics exists - so there is no need to involve a lambda capturing the object.

CookiePLMonster commented:
@dolphin-emu-bot rebuild

CookiePLMonster commented:
So with build machine now migrated to Visual Studio 2019, and build succeeding, any blockers preventing from merging this in?

CookiePLMonster commented:
Rebased on latest master and used the most up to date WIL. Doesn't bring any improvements to the PR but I think it makes sense to start off the most up to date submodule.

CookiePLMonster commented:
Pushed a tiny change, since I noticed `IID_PPV_ARGS` can be used in one more place.

goldenercobra commented on Source/Core/AudioCommon/WASAPIStream.cpp:
```diff
@@ -39,177 +45,123 @@ WASAPIStream::WASAPIStream()
   {
-    _com_error err(result);
-    std::string error = TStrToUTF8(err.ErrorMessage()).c_str();
+    std::string error;
```
minor optimization ([saves like 10 lines of assembly](https://godbolt.org/z/Lf3-u1)), but you can avoid reassigning a string by using an immediately invoked lambda 

```suggestion
    const std::string error = [&result] {
       switch (result)
       {
       case AUDCLNT_E_DEVICE_IN_USE:
           return "Audio endpoint already in use!";
       default:
           return TStrToUTF8(_com_error(result).ErrorMessage()).c_str();
       }
    }();
```

CookiePLMonster replied:
Eh, not a fan. This is a very cold code path (ideally is never taken) and the improvement is small enough not to warrant this change which IMO decreases readability.

It's a neat trick, but I don't think this is the place to use it.

goldenercobra replied:
totally fair, just a big fan of immediately invoked lambdas :)

CookiePLMonster replied:
Sure thing, I am using them too, mostly for static local variable initializations - but I also think this is not the place to go for it :)

CookiePLMonster commented:
@dolphin-emu-bot rebuild

CookiePLMonster commented:
Updated `HandleWinAPI`, since I just had a realization that I can pass the string size to `ERROR_LOG` (since it uses sprintf internally) and not have to promote `string_view` to `string` just to gain a guaranteed null terminator.

CookiePLMonster commented:
Rebased on latest master due to conflicts.

CookiePLMonster commented:
I've rebased against the latest master since I updated WIL to the newest version - makes sense to keep it up to date if it's to become a submodule.

However, since looks like this is not getting merged anytime soon and I **really** want to use WIL in other PR's, maybe I should split the commit adding WIL as submodule as a separate PR so I don't stall waiting for this to be merged?

@BhaaLseN @Warepire @lioncash  thoughts?

BhaaLseN commented:
Not sure how others are coming along with testing and their reviews; but if that still takes a while it probably isn't a bad idea to pull it out into its own PR.

CookiePLMonster commented:
Postponed until #8427 gets merged.

leoetlino commented:
#8427 has now been merged :)

CookiePLMonster commented:
Rebased on latest master. Thankfully, all I had to do was remove the commit added WIL since it's now merged in separately!

CookiePLMonster commented:
@dolphin-emu-bot rebuild

shuffle2 commented:
btw, thanks for adding WIL!
I just used it for https://github.com/dolphin-emu/dolphin/pull/9438/files , and I think it provides stuff (the [COM wrappers](https://github.com/microsoft/wil/wiki/WinRT-and-COM-wrappers), string wrappers, etc.) to make this PR even nicer, if you're willing to give it a shot...

CookiePLMonster commented:
> to make this PR even nicer, if you're willing to give it a shot...

Do you see any other places where WIL could be used in this PR? For COM pointers I personally prefer WRL because it is included in Windows headers, and so it is used here.

Filoppi commented:
Shame I had not seen this before starting my WASAPI works (nor anyone told me), though it's not very encouraging to see this still pending given over a year has passed and I have a lot more changes.
I think I've made a lot of the same improvements as you even if I haven't fully checked your code, but my rewrite is much more extensive (it allows any setting to be changed at runtime for example).
https://github.com/dolphin-emu/dolphin/pull/8983
While the WASAPI code there is finished, I still can't mark the PR as finished as I need to clean up other parts.

shuffle2 commented:
@Filoppi , let's leave this PR specifically for WASAPI resource management improvements.

Filoppi commented:
Never said to not keep this even if merging is gonna be annoying. I think the only improvement that mine doesn't have is converting to the new ComPtr. I should have fixed all memory leaks and resource management stuff, plus the deadlocks (which weren't exclusive to WASAPI).
Also changed the devices to be saved by ID and not name string because that would break if you had two devices with the same name, and multiplied the WASAPI volume by windows mixer volume so it's directly controllable by there as well, as it should be.
Anyway, this is great stuff, why hasn't it been merged :(.

CookiePLMonster commented:
Merge conflicts resolved. I decided against using `wil::CoCreateInstance` because:
1. it uses `wil::com_ptr` and I'm using `WRL::ComPtr`
2. It wouldn't improve code readability too much and we **do** make use of `CoCreateInstance`'s result value, which `wil::CoCreateInstance` doesn't expose easily

Adamillo commented:
Does this PR fix the dreaded audio issues on Wasapi?
